specs/design-black-box.md

```diff
@@ -72,10 +72,121 @@ specific ConfigMaps/Secrets.
+|   |-- app
+|   |   |-- apis       # contains data structures that describe Application
+|   |   `-- client     # contains client that provides Application CRUD operations
+|   `-- engine     # the engine implementation
+```
+
+The engine is responsible for reconciliation only which includes:
+- Interacting with the Kubernetes API: load and maintain the cluster state; pushing required changes to the Kubernetes API.
+- Reconciliation logic: based on provides target resource definition find which resources should be updated/deleted/created.
+- Syncing logic: determine the order in which resources should be modified; features like sync hooks, waves, etc.
+
+The manifests generation is out of scope and should be implemented by the Engine consumer.
+
+### Engine API
+
+The engine API includes Application data structure/client and `Engine` golang interface that allows configuring reconciliation process:
+
+**Engine interface** - provides set of that allows updating reconciliation settings and subscribe to engine events.
+```golang
+type Engine interface {
+	// Run starts reconciliation loop using specified number of processors for reconciliation and operation execution.
+	Run(ctx context.Context, statusProcessors int, operationProcessors int)
+
+	// SetReconciliationSettings updates reconciliation settings
+	SetReconciliationSettings(settings ReconciliationSettings)
+
+	// OnBeforeSync registers callback that is executed before each sync operation.
+	OnBeforeSync(callback func(appName string, tasks []SyncTaskInfo) ([]SyncTaskInfo, error)) Unsubscribe
+
+	// OnSyncCompleted registers callback that is executed after each sync operation.
+	OnSyncCompleted(callback func(appName string, state appv1.OperationState) error) Unsubscribe
+
+	// OnClusterCacheInitialized registers that that is executed when cluster cache initialization is completed.
+	OnClusterCacheInitialized(callback func(server string)) Unsubscribe
+
+	// OnResourceUpdated registers that that is executed when cluster resource got updated.
+	OnResourceUpdated(callback func(cluster string, un *unstructured.Unstructured)) Unsubscribe
+
+	// OnResourceRemoved registers that that is executed when cluster resource got removed.
+	OnResourceRemoved(callback func(cluster string, key kube.ResourceKey)) Unsubscribe
+
+	// OnAppEvent registers callback that is executed on every application event.
+	OnAppEvent(callback func(app *appv1.Application, info EventInfo, message string)) Unsubscribe
+}
+
+type Unsubscribe func()
+```
+
+**ReconciliationSettings data structure** - holds reconciliation settings
+```golang
+type ReconciliationSettings struct {
+	// AppInstanceLabelKey holds label key which is automatically added to every resource managed by the application
+	AppInstanceLabelKey string
+	// ResourcesFilter holds settigns which allows to configure list of managed resource APIs
+	ResourcesFilter resource.ResourcesFilter
+	// ResourceOverrides holds settings which customize resource diffing logic
```

Comment:
How does this work? what should be put in the keys?

Reply:
Key is `group/kind`. Here is example of real life configuration:


```
    certmanager.k8s.io/Certificate:
      # Lua script for customizing the health status assessment
      health.lua: |
        hs = {}
        if obj.status ~= nil then
          # some logic here
        end
        hs.status = "Progressing"
        hs.message = "Waiting for certificate"
        return hs
```

Do you think it makes sense to replace string with `scheme.GroupKind` ?

Reply:
done